tools/system_libs.py

```diff
@@ -1580,76 +1530,101 @@ def add_library(lib):
+    if shared.Settings.WASM_BACKEND:
+      add_library(system_libs_map['libc_rt_wasm'])
+    add_library(system_libs_map['libcompiler_rt'])
+  else:
+    # These libraris get included automatically based the symbols needed by the input file.
+    # Other system libraries are simply include by default.
+    for libname in ['libgl', 'libal', 'libhtml5', 'libembind']:
+      if libname in already_included:
+        continue
+      lib = system_libs_map[libname]
+      force_this = lib.name in force_include
+      if not force_this:
+        need_syms = set()
+        has_syms = set()
+        for symbols in symbolses:
+          if shared.Settings.VERBOSE:
+            logger.debug('undefs: ' + str(symbols.undefs))
+          for library_symbol in lib.symbols:
+            if library_symbol in symbols.undefs:
+              need_syms.add(library_symbol)
+            if library_symbol in symbols.defs:
+              has_syms.add(library_symbol)
+        for haz in has_syms:
+          if haz in need_syms:
+            # remove symbols that are supplied by another of the inputs
+            need_syms.remove(haz)
+        if shared.Settings.VERBOSE:
+          logger.debug('considering %s: we need %s and have %s' % (lib.name, str(need_syms), str(has_syms)))
+        if not len(need_syms):
+          continue
 
-  # the sanitizer runtimes may call mmap, which will need a few things. sadly
-  # the usual deps_info mechanism does not work since we scan only user files
-  # for things, and not libraries (to be able to scan libraries, we'd need to
-  # somehow figure out which of their object files will actually be linked in -
-  # but only lld knows that). so just directly handle that here.
-  if shared.Settings.UBSAN_RUNTIME or shared.Settings.USE_LSAN or shared.Settings.USE_ASAN:
-    shared.Settings.EXPORTED_FUNCTIONS.append(mangle_c_symbol_name('memset'))
+      # We need to build and link the library in
+      add_library(lib)
 
-  if shared.Settings.STANDALONE_WASM:
-    add_library(system_libs_map['libstandalonewasm'])
+    sanitize = shared.Settings.USE_LSAN or shared.Settings.USE_ASAN or shared.Settings.UBSAN_RUNTIME
 
-  if shared.Settings.PROXY_POSIX_SOCKETS:
-    add_library(system_libs_map['libsockets_proxy'])
-  else:
-    add_library(system_libs_map['libsockets'])
+    add_library(system_libs_map['libc'])
+    add_library(system_libs_map['libcompiler_rt'])
+    if not shared.Settings.WASM_BACKEND and not shared.Settings.MINIMAL_RUNTIME:
+      add_library(system_libs_map['libc-extras'])
+    if shared.Settings.WASM:
+      add_library(system_libs_map['libc-wasm'])
+    if cxx:
+      add_library(system_libs_map['libc++'])
+    if cxx or sanitize:
+      add_library(system_libs_map['libc++abi'])
+      if shared.Settings.EXCEPTION_HANDLING:
+        add_library(system_libs_map['libunwind'])
+    if shared.Settings.MALLOC != 'none':
+      add_library(system_libs_map['libmalloc'])
+    add_library(system_libs_map['libpthread'])
```

Comment:
The order here is important in some cases.  It took my a while to tune this change so that it would have minimal effect on the tests we run. 

If we want to tune the order as a followup we can.. and we can then also observe the effect on the tests results and decided if they are correct/desired.

Reply:
I wasn't aware of the importance of the order. Thanks!